src/v/transform/transform_processor.cc

```diff
@@ -0,0 +1,159 @@
+
+ss::future<> processor::start() {
+    try {
+        co_await _engine->start();
+        co_await _engine->initialize();
+    } catch (const std::exception& ex) {
+        vlog(_logger.warn, "error starting processor engine: {}", ex);
+        _error_callback(_id, _ntp, _meta);
+    }
+    _task = run_transform_loop();
+}
+
+ss::future<> processor::stop() {
+    _as.request_abort();
+    co_await std::exchange(_task, ss::now());
+    co_await _engine->stop();
+}
+
+ss::future<> processor::run_transform_loop() {
+    try {
+        co_await do_run_transform_loop();
+    } catch (const ss::abort_requested_exception&) {
+        // do nothing, we wanted to stop.
+    } catch (...) {
+        vlog(
+          _logger.warn,
+          "error running processor: {}",
+          std::current_exception());
+        _error_callback(_id, _ntp, _meta);
+    }
+}
+ss::future<> processor::do_run_transform_loop() {
+    auto offset = co_await _source->load_latest_offset();
+    vlog(_logger.trace, "starting at offset {}", offset);
+    // TODO(rockwood): Optimize this loop, we should not wait for the transform
+    // or for writing before attempting to read more data.
+    while (!_as.abort_requested()) {
+        auto reader = co_await _source->read_batch(offset, &_as);
+        auto latest_offset = co_await consume_batches(
+          std::move(reader).release(),
+          [this](model::record_batch_reader::data_t batches) {
+              return transform_batches(std::move(batches));
+          },
+          &_as);
+        if (!latest_offset.has_value() || latest_offset.value() <= offset) {
+            vlog(
+              _logger.trace,
+              "received no results, sleeping before polling at offset {}",
+              offset);
+            constexpr auto fallback_poll_interval = std::chrono::seconds(1);
+            simple_time_jitter<ss::lowres_clock> jitter(fallback_poll_interval);
+            try {
+                co_await ss::sleep_abortable<ss::lowres_clock>(
+                  jitter.next_duration(), _as);
+            } catch (const ss::sleep_aborted&) {
+                // do nothing, the next iteration of the loop will exit
+            }
+            continue;
+        }
+        offset = latest_offset.value();
+        vlog(_logger.trace, "consumed upto offset {}", offset);
+    }
+}
+
+ss::future<>
+processor::transform_batches(model::record_batch_reader::data_t batches) {
+    ss::chunked_fifo<model::record_batch> transformed_batches;
+    transformed_batches.reserve(batches.size());
+    for (auto& batch : batches) {
+        _probe->increment_read_bytes(batch.size_bytes());
+        auto transformed = co_await _engine->transform(
+          std::move(batch), _probe);
+        _probe->increment_write_bytes(transformed.size_bytes());
+        transformed_batches.push_back(std::move(transformed));
+    }
+    co_await _sinks[0]->write(std::move(transformed_batches));
```

Comment:
hmm, i don't see any place that checks if sinks is non-empty before this dereference

Reply:
It's an implicit assertion that you always 1 sink right now. I'll add a `vassert` in the constructor for now.